Add experimental torch.export prototype

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95070

This is WIP PR for adding torch.export API in OSS. Couple of points:
- I intentionally named it as experimental_export so that ppl don't get confused thinking this is our official API
- We don't plan to use AOTAutograd backend just yet. The reason we have it here is because the functionalization AOTAutograd uses is what we need for export (handling of param/buffer mutation etc). In the near future, I will extract the functionalization part and use it on top of make_fx. What we have right now is merely a placeholder. 
- The reason we want to do it now is because we want to have some minimal tests running in OSS so that we can catch regressions earlier. 

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

tugsbayasgalan committed:
Add experimental torch.export API

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add experimental torch.export API"




[ghstack-poisoned]

tugsbayasgalan commented on torch/_export/__init__.py:
```diff
@@ -0,0 +1,238 @@
+        fw_compiler=fw_compiler,
+        bw_compiler=lambda gm, inputs: None,
+        partition_fn=partition_fn,
+        decompositions=core_aten_decompositions(),
```
@SherlockNoMad  is this how we should use core_aten_decomp?

avikchaudhuri commented on torch/_export/__init__.py:
```diff
@@ -0,0 +1,238 @@
+    with using_config(DynamoConfig()):
+        graph_module, guards = torchdynamo.export(f, *args, aten_graph=False)
+    graph_module, _, out_spec = aot_capture(graph_module, flat_args)
+    return ExportedProgram(fw_module=graph_module, example_inputs=original_flat_args, in_spec=in_spec, out_spec=out_spec)
```
What about `guards`?

tugsbayasgalan replied:
For now, i just ignore them.

avikchaudhuri commented on torch/_export/__init__.py:
```diff
@@ -0,0 +1,238 @@
+
+    flattened_args, in_spec = pytree.tree_flatten(args)
+    original_flat_args = tuple(flattened_args)
+    flat_args = tuple(flattened_args)
```
same as `original_flat_args`?

tugsbayasgalan replied:
Yeah I did this to be extra safe about graph not modifying user input.

zhxchen17 commented on torch/_export/__init__.py:
```diff
@@ -0,0 +1,238 @@
+    full_args.extend(params_flat)
+    full_args.extend(flat_args)
+
+    assert any(arg.requires_grad for arg in full_args), "AOT Autograd needs at least one graph input to have .requires_grad=True, otherwise functionalization will be skipped. Please make sure at least one input has .requires_grad=True"
```
this won't be needed.

tugsbayasgalan replied:
removed

tugsbayasgalan committed:
Update on "[WIP] Add experimental torch.export API"


This is WIP PR for adding torch.export API in OSS. In the next PR, we need to address following:
1. Add __call__ method which unflattens/flattens input/output which matches eager semantics
2. Make sure the resulting graph is in compliant with Export IR. 
3. Add more good tests. 

[ghstack-poisoned]

tugsbayasgalan commented on torch/_export/__init__.py:
```diff
@@ -0,0 +1,238 @@
+        keep_inference_input_mutations=False,
+    )
+
+    @contextlib.contextmanager
+    def setup_dynamic_shape():
+        prev, torch._functorch.config.use_dynamic_shapes = (
+            torch._functorch.config.use_dynamic_shapes,
+            True,
+        )
+        try:
+            yield
+        finally:
+            torch._functorch.config.use_dynamic_shapes = prev
```
I think the default value can be different right? This context manager sets it to true unconditionally.

tugsbayasgalan commented on torch/_export/__init__.py:
```diff
@@ -0,0 +1,238 @@
+        torchdynamo.config.dynamic_shapes = self.dynamic_shapes
+        torchdynamo.config.specialize_int_float = self.specialize_int_float
+        torchdynamo.config.allow_rnn = self.allow_rnn
+        torchdynamo.config.verbose = self.verbose
```
Good idea. Removed the config. Tho i think it is still better to keep these configs around as default value in dynamo is still false.

yanboliang replied:
Echo to @zhxchen17, I think it should be a very thin layer at starting point, only includes the delta configs between ```export``` and normal dynamo.

tugsbayasgalan commented on torch/_export/__init__.py:
```diff
@@ -0,0 +1,238 @@
+    assert out_spec is not None
+    return graph_module, mutation, out_spec
+
+def export(f: Callable, args: Tuple, is_training=False):
```
changed

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "[WIP] Add experimental torch.export API"


This is WIP PR for adding torch.export API in OSS. In the next PR, we need to address following:
1. Add __call__ method which unflattens/flattens input/output which matches eager semantics
2. Make sure the resulting graph is in compliant with Export IR. 
3. Add more good tests. 

[ghstack-poisoned]

voznesenskym commented on torch/_export/__init__.py:
```diff
@@ -0,0 +1,228 @@
+
+
+@contextlib.contextmanager
+def _using_dynamo_config(**kwargs):
```
This is a weird util to have in an __init__.py

tugsbayasgalan replied:
Yeah i will remove this.

voznesenskym reviewed (changes requested):
I have a ton of questions about why we need this, and why we are doing this. A lot of this feels like it violates layering hierarchy, and provides APIs that only wrap existing APIs except imbued with use case specific assumptions.

voznesenskym commented on torch/_export/__init__.py:
```diff
@@ -0,0 +1,228 @@
+    In the current context we're just experimenting the idea so it's possible things
+    could break. For the next step we should find a way to upstream something reasonable.
+    """
+    param_list = [
+        *mod.named_parameters(remove_duplicate=False),
+        *mod.named_buffers(remove_duplicate=False),
+    ]
+    params = dict(param_list)
+    params_flat, params_spec = pytree.tree_flatten(params)
+    params_len = len(params_flat)
```
This is just taking code from aot_autograd.py and reimplementing it, why?

tugsbayasgalan replied:
Oh this is actually coming from Zhengxu's experiment on using AOTAutograd instead of make_fx. Since this change handles our use cases pretty well, this can serve as placeholder for our tests to pass as baseline. Soon, this will be replaced with functionalization (used in AOTAutograd) + make_fx.

zhxchen17 replied:
per @SherlockNoMad's advice we don't want to reuse aotautograd for now.

voznesenskym replied:
Pull into a util

gmagogsfm replied:
@tugsbayasgalan Shall we pull into a util so avoid forking code like Voz said?

jansel commented on torch/_export/__init__.py:
```diff
@@ -0,0 +1,228 @@
+@contextlib.contextmanager
+def _using_dynamo_config(**kwargs):
+    prev_configs : Dict[str, bool] = {}
+    for key in kwargs:
+        assert hasattr(torchdynamo.config, key)
+        prev_configs[key] = getattr(torchdynamo.config, key)
+
+    for key in kwargs:
+        setattr(torchdynamo.config, key, kwargs[key])
+    try:
+        yield
+    finally:
+        for key, value in prev_configs.items():
+            setattr(torchdynamo.config, key, value)
```
Duplicate of `torchdynamo.config.patch()`?

tugsbayasgalan replied:
oh oops. that sounds better

jansel commented on torch/_export/__init__.py:
```diff
@@ -0,0 +1,228 @@
+
+from .workflow import ExportedProgram
+
+CORE_ATEN_DECOMPOSITIONS_TABLE = core_aten_decompositions()
```
Why do we have a function if we are just constructing this in global scope?

tugsbayasgalan replied:
This is for caching purposes.

tugsbayasgalan committed:
Update on "[WIP] Add experimental torch.export API"


This is WIP PR for adding torch.export API in OSS. In the next PR, we need to address following:
1. Add __call__ method which unflattens/flattens input/output which matches eager semantics
2. Make sure the resulting graph is in compliant with Export IR. 
3. Add more good tests. 

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "[WIP] Add experimental torch.export API"


This is WIP PR for adding torch.export API in OSS. In the next PR, we need to address following:
1. Add __call__ method which unflattens/flattens input/output which matches eager semantics
2. Make sure the resulting graph is in compliant with Export IR. 
3. Add more good tests. 

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] Add experimental torch.export API"


This is WIP PR for adding torch.export API in OSS. Couple of points:
- I intentionally named it as experimental_export so that ppl don't get confused thinking this is our official API
- We don't plan to use AOTAutograd backend just yet. The reason we have it here is because the functionalization AOTAutograd uses is what we need for export (handling of param/buffer mutation etc). In the near future, I will extract the functionalization part and use it on top of make_fx. What we have right now is merely a placeholder. 
- The reason we want to do it now is because we want to have some minimal tests running in OSS so that we can catch regressions earlier. 

[ghstack-poisoned]

voznesenskym commented:
I think that this is a good chance to take a step back and revisit our export API. Does the dynamo one suffice? Should we unbundle make_fx/aot_autograd from .export and make the user do 2+ calls? What are our decomp notions? Do we ship with a canonical set as a variable? as a helper fn? etc. Let's maybe write up a quick UX overview here before jumping into an impl.

cc: @suo for export UX and API

zhxchen17 commented:
I think I'm okay with a prototype landed first because we need a place to iterate ideas without too much disconnection from pytorch OSS. Plus we call it experimental_export() so we don't mean to make it a public API any time soon.

voznesenskym commented:
> I think I'm okay with a prototype landed first because we need a place to iterate ideas without too much disconnection from pytorch OSS. Plus we call it experimental_export() so we don't mean to make it a public API any time soon.

If you want to experiment, just do it alongside the dynamo export api? Making it a torch level thing feels off to me. 

I am also not sure where there is to experiment on here? A lot of this is boilerplate on top of existing APIs. There's not like, a new algo we are trying out or anything, right?

zhxchen17 commented:
> > I think I'm okay with a prototype landed first because we need a place to iterate ideas without too much disconnection from pytorch OSS. Plus we call it experimental_export() so we don't mean to make it a public API any time soon.
> 
> If you want to experiment, just do it alongside the dynamo export api? Making it a torch level thing feels off to me.
> 
> I am also not sure where there is to experiment on here? A lot of this is boilerplate on top of existing APIs. There's not like, a new algo we are trying out or anything, right?

Correct me if I am wrong, but I think dynamo.export() shouldn't be a user interface exposed to users, but rather a low level interface to capture a FX graph out of the original PyTorch eager program. We want to be able to provide more information  to users in an organized way. The users here mean both PyTorch Edge team and some internal server use cases. Now if you ask "why we cannot replace all internal and Edge uses with dynamo.export()?", my answer would be "sure, as long as you will directly answer those people's questions, provide all the UX they want and address all their requests." 

So I think the question is not about whether torch.export() and dynamo.export() are essentially the same thing, it's more about how we divide the work into core mechanism and user support part. IMHO one serves as a semantic layer to users with all sorts of different use cases, and the other one serves as a pretty stable building block.

cc @suo I vaguely remember that's your initial idea right?

tugsbayasgalan commented:
@SherlockNoMad  is also experimenting with training support inside torchdynamo.export and I didn't want to collide with his work. If both of our code is inside torchdynamo.export, there can be many different number of configurations we may need to think about. Instead, if we use torchdynamo.export(aten_graph=False) as our stable building block and do some prototype work on top of it with some good unittests that export should handle, we can move forward without much friction. For example, when @SherlockNoMad makes changes to AOTAutograd or torchdynamo.export, we will be able to catch regressions easier in OSS.

gmagogsfm commented:
> Correct me if I am wrong, but I think dynamo.export() shouldn't be a user interface exposed to users, but rather a low level interface to capture a FX graph out of the original PyTorch eager program. 
> So I think the question is not about whether torch.export() and dynamo.export() are essentially the same thing

We have reached the conclusion that dynamo.export() isn't the user-facing API (torch.export) a while ago. To be more principled, we need a user-facing API and a building-block API that supports the former. Specifically, their responsibilities differ in following aspects:

1) Building-block API focuses on parsing Python execution semantics from user program, and turns them into an FX graph and a set of guards. User-facing API needs to take care of all other responsibilities, including supporting exporting module with multiple methods, guiding users to understand and remove guards. They do not necessarily have a tight coupling with building-block API's core value proposition, which is understanding Python.

2) User-facing API should hide complexity and evolution in underlying Building-block API. For example, different tracing-modes, decomposition table, various configs knobs, are implementation details that are subject to changes and should not be "directly" exposed to users. Instead, they should be hidden behind user-facing APIs.

3) User-facing API may absorb some workflow-specific logic, like features to be composable with a specific compilation flow. building-block API should shy away from the complexity there.

In today's world, I think it makes most sense for Dynamo.export to be "only" the building-block API instead of being both and absorbing unnecessary complexity that are less relevant to its core value proposition (understanding python). Therefore, I do not think adding export logic "along side" dynamo.export, ie inside Dynamo is correct. @suo @avikchaudhuri thoughts?


> If you want to experiment

I think "experiment" is a misnomer here. It means that this is "experimental", aka unofficial rather than an experiment is being conducted.  


@voznesenskym It seems you are uncomfortable exposing experimental functionality as torch.* level API, I am uncomfortable with it either. But this PR doesn't really expose a torch.* level API, instead it is adding an experimental API not in a somewhat hidden _export namespace. Does this clear things up a bit?

yanboliang commented:
Per my understanding after reading this thread, @tugsbayasgalan mentioned the major reason behind this PR at description part: 

> The reason we want to do it now is because we want to have some minimal tests running in OSS so that we can catch regressions earlier.

I think this is reasonable, if you don't have this experimental API, you won't know how it works and what it should look like. I do agree this shouldn't be a public API as we need careful design and more discussion on how it should be. But this can be a good starting point to iterate, it can also help us to collection more insights on how this public API should be.

SherlockNoMad commented:
I think these are the problems we will need to address for export path:
1. Existing dynamo.export() output graph doesn't comply with the export IR spec. 
   a. The output graph is not functional. We currently directly call make_fx without functionalization. We can fix this by reusing aot_autograd code, or manually invoke functionalization + make_fx.
   b. The output graph needs be a compliant Core Aten FX graph by default. Specifically for dynamo.export(), aten_graph should be True by default, and decomposition_table should use core_aten_decompositions. 
2. Need to decide on the UX on supporting both inference mode and training mode export. They doesn't necessarily need to share the same API, but we need some place as the development ground. Otherwise, we will need to cluster everything into dynamo.export() API, like it was done in https://github.com/pytorch/pytorch/pull/95023/files#diff-c663df72d7a648204ff256e604fe3c9b393e7e2d58204e3bc818d90a99712911L521-R522 
3. We need to figure out the guard resolution UX. 

ATM, dynamo.export() is used by exir.capture(pt2_mode=True), which is heavily used by the internal workflows. This make it practically difficult to modify dynamo.export(). I am also running into cases where OSS CI passed and Internal CI failed, and PRs got reverted. 

So I think we need a place to iterate, without breaking internal builds. The API should remain purely experimental, until we explicitly publish it for internal/public use.

voznesenskym commented:
> I think these are the problems we will need to address for export path:
> 
> 1. Existing dynamo.export() output graph doesn't comply with the export IR spec. 
> 
>    a. The output graph is not functional. We currently directly call make_fx without functionalization. We can fix this by reusing aot_autograd code, or manually invoke functionalization + make_fx.
> 
>    b. The output graph needs be a compliant Core Aten FX graph by default. Specifically for dynamo.export(), aten_graph should be True by default, and decomposition_table should use core_aten_decompositions. 
> 
> 2. Need to decide on the UX on supporting both inference mode and training mode export. They doesn't necessarily need to share the same API, but we need some place as the development ground. Otherwise, we will need to cluster everything into dynamo.export() API, like it was done in https://github.com/pytorch/pytorch/pull/95023/files#diff-c663df72d7a648204ff256e604fe3c9b393e7e2d58204e3bc818d90a99712911L521-R522 
> 
> 3. We need to figure out the guard resolution UX. 
> 
> 
> 
> ATM, dynamo.export() is used by exir.capture(pt2_mode=True), which is heavily used by the internal workflows. This make it practically difficult to modify dynamo.export(). I am also running into cases where OSS CI passed and Internal CI failed, and PRs got reverted. 
> 
> 
> 
> So I think we need a place to iterate, without breaking internal builds. The API should remain purely experimental, until we explicitly publish it for internal/public use. 

1) Put it dynamo/experimental to not break layering 

2) don't duplicate copy/paste code from aot_autograd or other places, pull it into utils

Unbundle more, you don't need one stop shop APIs, it's okay if the user has to call a few experimental functions to fulfill their full solution.

voznesenskym reviewed (changes requested):
See comment above, I think I am okay with an experimental API, but it needs a little more iteration.

zhxchen17 commented:
we will continue with our original plan

gmagogsfm commented:
Went through all the comments, there doesn't seem to be directional difference any more, only minor details. I will approve it now assuming Tugsuu will address those minor details.

tugsbayasgalan committed:
Update on "Add experimental torch.export prototype"


This is WIP PR for adding torch.export API in OSS. Couple of points:
- I intentionally named it as experimental_export so that ppl don't get confused thinking this is our official API
- We don't plan to use AOTAutograd backend just yet. The reason we have it here is because the functionalization AOTAutograd uses is what we need for export (handling of param/buffer mutation etc). In the near future, I will extract the functionalization part and use it on top of make_fx. What we have right now is merely a placeholder. 
- The reason we want to do it now is because we want to have some minimal tests running in OSS so that we can catch regressions earlier. 

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add experimental torch.export prototype"


This is WIP PR for adding torch.export API in OSS. Couple of points:
- I intentionally named it as experimental_export so that ppl don't get confused thinking this is our official API
- We don't plan to use AOTAutograd backend just yet. The reason we have it here is because the functionalization AOTAutograd uses is what we need for export (handling of param/buffer mutation etc). In the near future, I will extract the functionalization part and use it on top of make_fx. What we have right now is merely a placeholder. 
- The reason we want to do it now is because we want to have some minimal tests running in OSS so that we can catch regressions earlier. 

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge -f "Unrelated test failures"
